network/router.go

```diff
@@ -0,0 +1,343 @@
+				// remote connection closed
+				log.Lvl3(r.address, "handleConn with closed connection: stop (dst=", remote.Address, ")")
+				return
+			}
+			// weird error let's try again
+			log.Lvl3(r.id, "Error with connection", address, "=>", err)
+			continue
+		}
+
+		if err := r.Dispatch(&packet); err != nil {
+			log.Lvl3("Error dispatching:", err)
+		}
+	}
+}
+
+// connection returns the connection associated with this ServerIdentity. Nil if
+// nothing found. It always return the first connection associated.
+func (r *Router) connection(sid ServerIdentityID) Conn {
+	r.connsMut.Lock()
+	defer r.connsMut.Unlock()
+	arr := r.connections[sid]
+	if len(arr) == 0 {
+		return nil
+	}
+	return arr[0]
+}
+
+// registerConnection registers an ServerIdentity for a new connection, mapped with the
+// real physical address of the connection and the connection itself
+// it locks (and unlocks when done):  networkLock
+func (r *Router) registerConnection(remote *ServerIdentity, c Conn) {
+	log.Lvl4(r.address, "Registers", remote.Address)
+	r.connsMut.Lock()
+	defer r.connsMut.Unlock()
+	_, okc := r.connections[remote.ID]
+	if okc {
+		log.Lvl5("Connection already registered. Appending new connection to same identity.")
+	}
+	r.connections[remote.ID] = append(r.connections[remote.ID], c)
+}
+
+func (r *Router) launchHandleRoutine(dst *ServerIdentity, c Conn) {
+	r.wg.Add(1)
+	go r.handleConn(dst, c)
+}
+
+// Close shuts down all network connections and returns once all processing go
+// routines are done.
+func (r *Router) stopHandling() error {
+	return nil
+}
+
+// Closed returns true if the router is closed (or is closing). For a router
+// to be closed means that a call to Stop() must have been made.
+func (r *Router) Closed() bool {
+	r.closedMut.Lock()
+	defer r.closedMut.Unlock()
+	return r.isClosed
+}
+
+// close set the isClosed variable to true, any subsequent call to Closed()
+// will return true.
+func (r *Router) close() {
+}
+
+// Tx implements monitor/CounterIO
+// It returns the Tx for all connections managed by this router
+func (r *Router) Tx() uint64 {
+	r.connsMut.Lock()
+	defer r.connsMut.Unlock()
+	var tx uint64
+	for _, arr := range r.connections {
+		for _, c := range arr {
+			tx += c.Tx()
+		}
+	}
+	return tx
+}
+
+// Rx implements monitor/CounterIO
+// It returns the Rx for all connections managed by this router
+func (r *Router) Rx() uint64 {
+	r.connsMut.Lock()
+	defer r.connsMut.Unlock()
+	var rx uint64
+	for _, arr := range r.connections {
+		for _, c := range arr {
+			rx += c.Rx()
+		}
+	}
+	return rx
+}
+
+// Listening returns true if this router is listening or not.
+func (r *Router) Listening() bool {
+	return r.host.Listening()
+}
+
+// receiveServerIdentity takes a fresh new conn issued by the listener and
+// wait for the server identities of the remote party. It returns
+// the ServerIdentity of the remote party and register the connection.
+func (r *Router) receiveServerIdentity(c Conn) (*ServerIdentity, error) {
```

Comment:
Yeah I think it's a bit too big, so I kept it, but I removed the `SendServerIdentity`.